### PR TITLE
Compositional metrics

### LINKED ISSUE #5392: [Metrics] Add Depth Loss
## 🚀 Feature

Add depth-loss to PT-lighting since image-gradient is being added. As discussed [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/5056#issuecomment-745100976).

### Motivation

Depth-Loss is one of the metrics that's often used in depth-estimation-oriented papers (DenseDepth etc). The addition of depth-loss to PT-lightning out-of-box would greatly help developers and researchers

### Pitch

Make depth-loss available out-of-box in PT-Lightning

### Additional context

This is a continuation of PR #5056

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Fixes #5392 

Implements composition of metrics with simple operator interface like `metric_a + metric_b`.
## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by teddykoker** [comment]
Very cool! Is there a reason that the `CompositionalMetric` import is need for each operation?

Also, is there a way we could register all of these methods dynamically? Something like:
```python

operations = {
    "__add__": torch.add,
    "__mul__": torch.mul,
    ...
}

def register_operations():
    for attr, fn in operations.items():
        setattr(Metric, attr,  lambda self, other: CompositionalMetric(fn, self, other))
```

**Comment by justusschock** [comment]
@teddykoker the registering should be possible, we just have to split it into 2 groups: The ones that have two arguments and the ones having only one arg. I also think, that we should not use lambdas here since that way we might run into issues for some DP/DDP backends due to pickling.

regarding the import: it has to be somewhat locally, since the compositional metric depends on `Metric` and can therefore not yet be imported before the definition of `Metric` itself. this is why I now imported it locally so that the import is only triggered, when the function is called. also the import is written there multiple times, but essentially only done once, since python caches imports

**Comment by SkafteNicki** [comment]
Since some of the operations also need the input in the opposite order i.e `__radd__` uses `CompositionalMetric(torch.add, other, self)`, IMO it would probably be better just to leave it as it is (hopefully this should not need to be changed in the future)

**Comment by justusschock** [comment]
@SkafteNicki I have one more question for you. I just reused your kwargs filter, but right now I noticed, that if the user does not explicitly define argument names but just `*args, **kwargs` the filter does not work, right? Should we in that case forward all the kwargs?

**Comment by SkafteNicki** [comment]
You are right that it does not work, since the signature keys would then be simply `args` and `kwargs`
``` python
import inspect
def f(*args, **kwargs):
    return 1
signature = inspect.signature(f)
print(signature.parameters.keys()) #odict_keys(['args', 'kwargs'])
``` 
I agree that in this case we should just forward everything.

File: docs/source/metrics.rst [hunk 1]
```diff
@@ -250,6 +250,50 @@ In practise this means that:
...
+******************
+Metric Arithmetics
+******************
+
+Metrics support most of python built-in operators for arithmetic, logic and bitwise operations.
+
+For example for a metric that should return the sum of two different metrics, implementing a new metric is an overhead that is not necessary. 
+It can now be done with:
+
+.. code-block:: python
```
Comment (SkafteNicki) [inline docs/source/metrics.rst:262]: add empty line to fix doctest
```suggestion
.. code-block:: python

```

File: pytorch_lightning/metrics/compositional.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            return self.op(val_a)
+
+        return self.op(val_a, val_b)
+
+    def reset(self):
+        if isinstance(self.metric_a, Metric):
+            self.metric_a.reset()
+
+        if isinstance(self.metric_b, Metric):
+            self.metric_b.reset()
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/compositional.py:77]: just tried messing around locally and observed that this function also needs to be implemented
```suggestion
            self.metric_b.reset()
            
    def persistent(self, mode: bool = False):
        if isinstance(self.metric_a, Metric):
            self.metric_a.persistent(mode=mode)
        if isinstance(self.metric_b, Metric):
            self.metric_b.persistent(mode=mode)
```

**Comment by SkafteNicki** [comment]
@justusschock can we somehow add the operation to the output print?
If I run the following:
``` python
from pytorch_lightning.metrics import MeanSquaredError
m = MeanSquaredError()
print(m * m + m)
```
i get the following output:
``` python
CompositionalMetric(
  (metric_a): CompositionalMetric(
    (metric_a): MeanSquaredError()
    (metric_b): MeanSquaredError()
  )
  (metric_b): MeanSquaredError()
)
```
it would be nice if the print contained the operations that went into creating the compositional metric, so it is possible to backtrack how it was created.

**Comment by justusschock** [comment]
@SkafteNicki Done

**Review by SkafteNicki — APPROVED** [review]
small fix, else LGTM

**Review by teddykoker — APPROVED** [review]
@justusschock  if we put the two modules in the same file could we avoid the import issue?

File: pytorch_lightning/metrics/metric.py [hunk 3]
```diff
@@ -302,6 +297,199 @@ def state_dict(self, *args, **kwargs):
...
+    def __hash__(self):
+        hash_vals = [self.__class__.__name__]
+
+        for key in self._defaults.keys():
+            hash_vals.append(getattr(self, key))
+
+        return hash(tuple(hash_vals))
+
+    def __add__(self, other: Any):
+        from pytorch_lightning.metrics.compositional import CompositionalMetric
```
Comment (Borda) [inline pytorch_lightning/metrics/metric.py:326]: why do we need to import it here? bc cyclic import?

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/metrics/metric.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py:326]: exactly

**Comment by justusschock** [comment]
@teddykoker I'm not sure about that. Because we still have cyclic dependencies.

**Comment by SkafteNicki** [comment]
@justusschock just tried locally, there should be no problems with moving `CompositionalMetric` to `metric.py` and still have it working

File: docs/source/metrics.rst [hunk 4]
```diff
@@ -250,6 +250,51 @@ In practise this means that:
...
+Similarly ``new_metric.compute()`` now calls compute of ``first_metric`` and ``second_metric`` and adds the results up.
+
+This pattern is implemented for the following operators (with ``a`` being metrics and ``b`` being metrics, tensors, integer or floats):
+
+* Addition (``a + b``)
+* Bitwise AND (``a & b``)
+* Equality (``a == b``)
+* Floordivision (``a // b``)
+* Greater Equal (``a >= b``)
+* Greater (``a >b ``)
```
Comment (SkafteNicki) [inline docs/source/metrics.rst:281]: docs are failing on this line with: "Inline literal start-string without end-string." Not sure if this fixes it
```suggestion
* Greater (``a > b``)
```

**Comment by justusschock** [comment]
@teddykoker do you want to move it then?

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@justusschock it seems that we have merged something where most tests are failing https://github.com/PyTorchLightning/pytorch-lightning/runs/1769896518
could you please check it...
